File: voice-note-server/src/main/kotlin/com/suikouji/server/llm/dto/TransactionCorrectionRequest.kt
package com.suikouji.server.llm.dto

import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.Size

data class TransactionCorrectionRequest(
    @field:NotEmpty
    @field:Size(max = 10, message = "Batch cannot exceed 10 items")
    @field:Valid
    val currentBatch: List<BatchItem>,

    @field:NotBlank
    @field:Size(max = 2000, message = "Correction text must be at most 2000 characters")
    val correctionText: String,

    @field:Valid
    val context: ParseContext? = null
)

data class BatchItem(
    val index: Int,
    val amount: Double?,
    @field:Size(max = 100, message = "Category must be at most 100 characters")
    val category: String?,
    @field:Size(max = 20, message = "Type must be at most 20 characters")
    val type: String?,
    @field:Size(max = 500, message = "Description must be at most 500 characters")
    val description: String?,
    @field:Size(max = 30, message = "Date must be at most 30 characters")
    val date: String? = null
)

File: voice-note-client/lib/core/network/dto/transaction_correction_request.dart
import 'transaction_parse_request.dart';

/// Request body for POST /api/v1/llm/correct-transaction.
class TransactionCorrectionRequest {
  final List<BatchItem> currentBatch;
  final String correctionText;
  final ParseContext? context;

  const TransactionCorrectionRequest({
    required this.currentBatch,
    required this.correctionText,
    this.context,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'currentBatch': currentBatch.map((e) => e.toJson()).toList(),
      'correctionText': correctionText,
    };
    if (context != null) json['context'] = context!.toJson();
    return json;
  }
}

/// A single item in the correction batch context.
class BatchItem {
  final int index;
  final double? amount;
  final String? category;
  final String? type;
  final String? description;
  final String? date;

  const BatchItem({
    required this.index,
    this.amount,
    this.category,
    this.type,
    this.description,
    this.date,
  });

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'index': index,
      if (amount != null) 'amount': amount,
      if (category != null) 'category': category,
      if (type != null) 'type': type,
      if (description != null) 'description': description,
      if (date != null) 'date': date,
    };
  }
}
